File: examples/k6/conference.js
/**
 * Conference Bridge Load Test
 *
 * Each VU dials into a conference bridge URI with N participant legs.
 * The bridge (Asterisk ConfBridge / FreeSWITCH) mixes audio.
 * xk6-sip-media manages N independent SIP/RTP legs per VU.
 *
 * Usage:
 *   SIP_BRIDGE="sip:conf-room-101@pbx" SIP_PARTICIPANTS=3 \
 *     ./k6 run examples/k6/conference.js
 */
import sip from 'k6/x/sip';
import { check, sleep } from 'k6';

const BRIDGE       = __ENV.SIP_BRIDGE        || 'sip:conf101@192.168.1.100';
const PARTICIPANTS = parseInt(__ENV.SIP_PARTICIPANTS || '3', 10);
const DURATION     = __ENV.SIP_DURATION      || '30s';

export const options = {
  scenarios: {
    conference_load: {
      executor:    'ramping-vus',
      startVUs:    0,
      stages: [
        { duration: '30s', target: 10 },  // 10 VUs × 3 legs = 30 conference legs
        { duration: '60s', target: 30 },  // 30 VUs × 3 legs = 90 conference legs
        { duration: '30s', target: 0  },
      ],
    },
  },
  thresholds: {
    sip_call_success:    ['count>0'],
    sip_conference_legs: ['avg>=1'],
    mos_score:           ['avg>=3.5'],
    rtp_jitter_ms:       ['avg<80'],  // bridges can add extra buffering
  },
};

export default function () {
  // Start conference with host leg (dial 1)
  const conf = sip.conference({
    bridgeURI: BRIDGE,
    audio:     { file: './examples/audio/sample.wav' },
    duration:  DURATION,
  });

  // Add remaining participants (dials 2…N)
  for (let i = 1; i < PARTICIPANTS; i++) {
    const err = conf.addParticipant('', { // empty = use same bridgeURI
      audio: { file: './examples/audio/sample.wav' },
    });
    if (err !== null) {
      console.warn(`participant ${i + 1} failed to join: ${err}`);
    }
  }

  check(conf, {
    'all participants joined': (c) => c.participantCount() === PARTICIPANTS,
  });

  // Conference runs until duration expires (set on each leg)
  conf.waitDone();

  const r = conf.result();
  check(r, {
    'conference completed':   (r) => r.legs > 0,
    'avg MOS >= 3.5':         (r) => r.avg_mos >= 3.5,
    'packet loss < 5%':       (r) => r.packets_lost / Math.max(r.packets_sent, 1) < 0.05,
  });

  console.log(
    `conf done: legs=${r.legs} avg_mos=${r.avg_mos.toFixed(2)} ` +
    `min_mos=${r.min_mos.toFixed(2)} lost=${r.packets_lost}/${r.packets_sent}`,
  );
}

File: sip/conference.go
package sip

import (
	"fmt"
	"math/rand"
	"sync"
	"time"

	corertp "xk6-sip-media/core/rtp"
)

// ConferenceConfig describes a bridge-based SIP conference.
//
// Each participant is dialed as an independent SIP call leg to BridgeURI.
// The conference bridge (Asterisk ConfBridge / FreeSWITCH) mixes audio.
// No local audio mixing is performed by xk6-sip-media.
type ConferenceConfig struct {
	// BridgeURI is the conference bridge SIP URI, e.g. "sip:room101@pbx".
	BridgeURI string

	// AudioFile is the WAV file streamed by each participant leg.
	AudioFile string

	// Codec is the audio codec for all legs (default PCMU).
	Codec string

	// LocalIP for all legs (default: auto-detect).
	LocalIP string

	// Duration caps each leg. 0 = manual Hangup() required.
	Duration time.Duration

	// Username / Password for digest auth on each INVITE (optional).
	Username string
	Password string
}

// Conference manages a set of concurrent SIP call legs all connected to the
// same conference bridge URI.
type Conference struct {
	mu   sync.Mutex
	legs []*CallHandle
	cfg  ConferenceConfig
}

// ConferenceResult aggregates quality metrics across all legs.
type ConferenceResult struct {
	Legs             int
	AvgMOS           float64
	MinMOS           float64
	MaxMOS           float64
	TotalPacketsSent int64
	TotalPacketsLost int64
}

// StartConference dials BridgeURI once (as the "host" leg) and returns
// a *Conference handle. Use AddParticipant to add more legs.
func StartConference(cfg ConferenceConfig) (*Conference, error) {
	c := &Conference{cfg: cfg}

	// Dial the initial host leg to open the bridge room
	leg, err := c.dialLeg(cfg.BridgeURI, nil)
	if err != nil {
		return nil, fmt.Errorf("conference: dial host leg: %w", err)
	}

	c.mu.Lock()
	c.legs = append(c.legs, leg)
	c.mu.Unlock()

	return c, nil
}

// AddParticipant dials targetURI (or BridgeURI if empty) into the conference.
// Each leg gets a unique RTP port so concurrent participants do not collide.
// If cfg is non-nil, its AudioFile, Codec, and LocalIP fields override the
// conference-level defaults for this leg only.
func (c *Conference) AddParticipant(targetURI string, cfg *ConferenceConfig) error {
	uri := targetURI
	if uri == "" {
		uri = c.cfg.BridgeURI
	}
	leg, err := c.dialLeg(uri, cfg)
	if err != nil {
		return fmt.Errorf("conference add participant %s: %w", uri, err)
	}

	c.mu.Lock()
	c.legs = append(c.legs, leg)
	c.mu.Unlock()
	return nil
}

// ParticipantCount returns the number of active legs.
func (c *Conference) ParticipantCount() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	active := 0
	for _, l := range c.legs {
		if l.IsActive() {
			active++
		}
	}
	return active
}

// Hangup terminates all active legs.
func (c *Conference) Hangup() error {
	c.mu.Lock()
	legs := make([]*CallHandle, len(c.legs))
	copy(legs, c.legs)
	c.mu.Unlock()

	var lastErr error
	for _, leg := range legs {
		if err := leg.Hangup(); err != nil {
			lastErr = err
		}
	}
	return lastErr
}

// WaitDone blocks until all conference legs have ended.
func (c *Conference) WaitDone() {
	c.mu.Lock()
	legs := make([]*CallHandle, len(c.legs))
	copy(legs, c.legs)
	c.mu.Unlock()

	for _, leg := range legs {
		leg.WaitDone()
	}
}

// Result aggregates quality metrics across all legs once the conference ends.
func (c *Conference) Result() ConferenceResult {
	c.mu.Lock()
	legs := make([]*CallHandle, len(c.legs))
	copy(legs, c.legs)
	c.mu.Unlock()

	res := ConferenceResult{
		Legs:   len(legs),
		MinMOS: 5.0,
	}

	for _, leg := range legs {
		r := leg.Result() // blocks per leg until done
		res.TotalPacketsSent += int64(r.PacketsSent)
		res.TotalPacketsLost += int64(r.PacketsLost)
		res.AvgMOS += r.MOS
		if r.MOS < res.MinMOS {
			res.MinMOS = r.MOS
		}
		if r.MOS > res.MaxMOS {
			res.MaxMOS = r.MOS
		}
	}
	if len(legs) > 0 {
		res.AvgMOS /= float64(len(legs))
	}
	return res
}

// dialLeg creates one CallHandle for a single conference participant leg.
// If override is non-nil, its AudioFile, Codec, and LocalIP fields take
// precedence over the conference-level config for this leg.
func (c *Conference) dialLeg(targetURI string, override *ConferenceConfig) (*CallHandle, error) {
	// Each leg gets a unique From user to simulate different participants
	fromUser := fmt.Sprintf("conf-leg-%d", rand.Intn(999999))
	_ = fromUser // future: set custom From header

	cfg := CallConfig{
		Target:    targetURI,
		AudioFile: c.cfg.AudioFile,
		Codec:     c.cfg.Codec,
		LocalIP:   c.cfg.LocalIP,
		Duration:  c.cfg.Duration,
		RTPPort:   0, // auto-assign unique port per leg
	}
	if override != nil {
		if override.AudioFile != "" {
			cfg.AudioFile = override.AudioFile
		}
		if override.Codec != "" {
			cfg.Codec = override.Codec
		}
		if override.LocalIP != "" {
			cfg.LocalIP = override.LocalIP
		}
	}
	if cfg.Codec == "" {
		cfg.Codec = "PCMU"
	}

	return Dial(cfg)
}

// Compile-time check: ConferenceResult embeds no external types.
var _ = corertp.CallResult{}
